### PR TITLE
Add extended-monitor spec

### PR DESCRIPTION
**Pull request by emersion**
Closes: https://github.com/ircv3/ircv3-specifications/issues/79

### CODE REVIEW

**Comment by ProgVal** [comment]
Would it make sense to add `NICK` changes too?

**Comment by emersion** [comment]
MONITOR explicitly avoids sending NICK updates:

> The MONITOR implementation also enhances user privacy by disallowing subscription to hostmasks, allowing users to avoid nick-change stalking.

**Comment by ProgVal** [comment]
`account-notify` and `chghost` may have similar issues then; as they are likely to contain the new nickname.

**Comment by SadieCat** [comment]
If someone changes nickname it is typical behaviour to mark them as offline. If a user is marked as offline then they shouldn't receive anything enabled by this cap.

**Comment by delthas** [comment]
> account-notify and chghost may have similar issues then; as they are likely to contain the new nickname.
CHGHOST and SETNAME contain the current nick. If the nick is changed, then we won't receive any more CHGHOST or SETNAME messages that would contain the new nick (because MONITOR wouldn't track the new nick). Or am I missing something?

**Comment by ProgVal** [comment]
Not if they change host and name before changing nick.

**Comment by emersion** [comment]
>      Do we want to keep these semantics for the specs this adds?

I don't think so, since none of the specs (away-notify, account-notify, chghost, setname) specify that the server must send the initial state on JOIN. Instead, clients are expected to query the initial state with WHO/WHOX.

**Comment by SadieCat** [comment]
>Do we want to keep these semantics for the specs this adds?

I can see how this would be useful for clients. It would allow you to monitor someone when opening a query window (something some clients already do) and immediately get their relevant info.

**Comment by emersion** [comment]
Same applies to JOIN messages, without MONITOR. When another user joins a channel, clients want to immediately get relevant info for that user, to e.g. display the username differently in the list if it's away, among other things.

**Comment by SadieCat** [comment]
JOIN is slightly different imo as:

1. It's not feasible to send a ton of messages to users on join because of flood queues. Requiring clients that want that information to use WHO scales better.
2. Immediately having the account name, away message, etc is of less interest for channels than private messages.

**Comment by emersion** [comment]
I see. So basically sending the initial metadata is not technically reasonable for JOIN, but may fine be for MONITOR.

IMHO, I'd still prefer not to do it: rhe cost of the inconsistency between the specs outweights the minor annoyance of having to send WHO/WHOX.

**Comment by k4bek4be** [comment]
Including account names is not strictly necessary in the case of private conversation start, because it's already communicated in a message tag. It won't hurt though, and maybe there are clients that do not want to implement tags.

**Comment by delthas** [comment]
Thinking about this again, a modern client that wants to track user state would want the user away status, but also their realname. In which case we'd have to do a WHO anyways. So sending the AWAY state on MONITOR + wouldn't be so useful b/c we'd send out a WHO right after anyways. @SadieCat thoughts?

**Comment by SadieCat** [comment]
@delthas WHO doesn't include the away message.

**Comment by delthas** [comment]
From the implementation of extended-monitor in the only 3 popular, maintained server that support MONITOR (see linked MRs), it follows that implementing extended-monitor can be relatively easy and non intrusive, if we don't send the initial state (AWAY, ACCOUNT, CHGHOST, SETNAME) on MONITOR +. Otherwise we'd have to regenerate these messages in the server somehow from the MONITOR + handler, but this would need intrusive changes to the servers. Instead, adding a function for sending a message to all monitors of a nick and calling that function when a user away, account, chghost, setname is sent is much simpler and works well.

Modern clients that track the user state will want the user away status, message, realname, and such. Today these clients already send a WHOIS and have handlers for the various WHOIS replies. If we wanted to replace that with all state being pushed automatically, these clients would have to add support for chghost, and a handler for chghost, in order to receive the user real name. I expect that clients would find it easier to just keep sending WHOIS as they did before.

So TL;DR:
- much easier for servers to implement extended-monitor if we keep it simple, not sending the initial away/account/chghost/setname state
- clients will still send a WHOIS when first getting the state, but it's fine, they already did it. otherwise they'd have to support chghost to get the realname. even then, it's a small point compared to server implementations.

**Comment by emersion** [comment]
Ping: any more feedback on this spec?

**Comment by progval** [comment]
I don't think this was addressed: https://github.com/ircv3/ircv3-specifications/pull/466#issuecomment-885005724

**Comment by emersion** [comment]
I don't understand, can you rephrase? Is your concern that there is an information leak when the user changes their host or realname before changing their nick?

If that's your concern, do you want to add a paragraph to recommend clients to change their nick before their host or realname?

**Comment by progval** [comment]
If they change host and realname after changing nick, then it will show up for people monitoring the new nick (although that's still likely).

It should at least be mentioned that the only way to avoid leaks is to change the info while reconnecting

**Comment by emersion** [comment]
But if someone monitors both the old and new nicks, reconnecting won't help either? Once the target user re-connects with the new nick, the monitoring user will receive `RPL_MONONLINE` and will able to query the new host/realname?

Note, extended-monitor doesn't make this scenario worse. MONITOR alone allows polling the host/realname.

**Comment by emersion** [comment]
Added a section about privacy considerations.

**Comment by emersion** [comment]
Ping: any more feedback on this spec?

**Comment by jwheare** [comment]
This seems like it has decent consensus and can be merged. Will do that next week if no further comments. Ping me if I forget.

**Comment by jwheare** [comment]
I just remembered this discussion https://github.com/ircv3/ircv3-ideas/issues/78

I'm a bit hesitant to merge new drafts until we have a decision on that. I'll try to resolve that first.

**Comment by jwheare** [comment]
Dropping draft/ isn't clear cut so unblocking this.
